README.md
# homebridge-airport-express-playing

[![mit license](https://badgen.net/badge/license/MIT/red)](https://github.com/apexad/homebridge-airport-express-playing/blob/master/LICENSE)
[![npm](https://img.shields.io/npm/v/homebridge-airport-express-playing)](https://www.npmjs.com/package/homebridge-airport-express-playing)
[![npm](https://badgen.net/npm/dt/homebridge-airport-express-playing)](https://www.npmjs.com/package/homebridge-airport-express-playing)
[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)
[![donate](https://badgen.net/badge/donate/paypal/91BE09)](https://www.paypal.me/apexadm)

[Homebridge](https://github.com/homebridge/homebridge) plugin that creates a smart speaker accessory that shows playing/paused for airport express devices and has optional switch that can be used to trigger home automations.

## Configuration
This easiest way to use this plugin is to use [homebridge-config-ui-x](https://www.npmjs.com/package/homebridge-config-ui-x).  
To configure manually, add to the `platforms` section of Homebridge's `config.json` after installing the plugin.

**Config:**
```json
{
  "platform": "AirportExpressPlaying",
  "name": "Airport Express Playing Platform",
  "showSwitch": true
}
```

## Notes
1. Will not work when AirPlaying from iTunes running on Win10
2. Will not work in many situations involving grouped accessories due to limitations in mDNS

## Sponsors
[mbmccormick](https://github.com/mbmccormick)  
[jsiegenthaler](https://github.com/jsiegenthaler) - orginal idea to use airport express txt record [posted to reddit](https://www.reddit.com/r/homebridge/comments/jxt9le/added_a_switch_in_homebridge_to_show_if_airport/)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.1](https://github.com/apexad/homebridge-airport-express-playing/compare/v2.0.0...v2.0.1) (2021-04-08)


### Bug Fixes

* stop mdns browser as soon as possible, timeout is last resort ([c3c611e](https://github.com/apexad/homebridge-airport-express-playing/commit/c3c611ef1bcc89ce1d602a3bae7b360f877eda9b))

## [2.0.0](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.9...v2.0.0) (2021-04-08)


### ⚠ BREAKING CHANGES

* acessories will reset 1 last time and break automations, but should be cached after that

* switched plugin to be a dynamic platform (from static) so accesories will persist on homebridge restarts, fixes [#8](https://github.com/apexad/homebridge-airport-express-playing/issues/8) ([5c9c442](https://github.com/apexad/homebridge-airport-express-playing/commit/5c9c4428d0b6fb81026ed8b98bf9f3f5166c9667))

### [1.4.9](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.8...v1.4.9) (2021-02-22)


### Bug Fixes

* clarified use of mDNS data in description/README ([33d5096](https://github.com/apexad/homebridge-airport-express-playing/commit/33d5096600585571d30323f082bddcb08b02d5b7))
* correcrt typo in logs ([8fff343](https://github.com/apexad/homebridge-airport-express-playing/commit/8fff343c60684b6a26ecc641cc0ef796a366c08c))

### [1.4.8](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.7...v1.4.8) (2021-01-14)


### Bug Fixes

* move try..catch into mdns update function ([acee65e](https://github.com/apexad/homebridge-airport-express-playing/commit/acee65e19a03661537f90509663c062102dd4cfc))

### [1.4.7](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.6...v1.4.7) (2021-01-14)


### Bug Fixes

* wrap mDNS calls in try...catch fixes [#7](https://github.com/apexad/homebridge-airport-express-playing/issues/7) ([4f4565b](https://github.com/apexad/homebridge-airport-express-playing/commit/4f4565bff8019f22509ddbe81102eb86d6766fdb))

### [1.4.5](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.4...v1.4.5) (2020-12-16)

### [1.4.4](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.3...v1.4.4) (2020-12-09)


### Bug Fixes

* change manufacturer to be clear this is a homebridge plugin and not an Apple device ([cf786f7](https://github.com/apexad/homebridge-airport-express-playing/commit/cf786f74d8464e5a7f3c7bb407b439143f30a1ca))

### [1.4.3](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.2...v1.4.3) (2020-12-09)


### Bug Fixes

* remove broken test file ([3087e09](https://github.com/apexad/homebridge-airport-express-playing/commit/3087e09911a1300f59e61923bf698e287de12302))
* use STOPPED if Airport Express not connected/Playing, INTERRUPTED if error ([6b1c32a](https://github.com/apexad/homebridge-airport-express-playing/commit/6b1c32a1c17f87b568e0ea6235ac1db5a27df229))

### [1.4.2](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.1...v1.4.2) (2020-12-08)


### Bug Fixes

* correction to bit 11 (playing status bit) ([ea3f1e7](https://github.com/apexad/homebridge-airport-express-playing/commit/ea3f1e789f870e853a9db08a03bd8f70d6598e63))

### [1.4.1](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.0...v1.4.1) (2020-12-07)


### Bug Fixes

* showSwitch value check corrected ([745630d](https://github.com/apexad/homebridge-airport-express-playing/commit/745630deb786de362647c674a37c0f7b713a1c09))

## [1.4.0](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.4.0-beta.3...v1.4.0) (2020-12-07)


### Features

* make switch optional with config option ([66475cb](https://github.com/apexad/homebridge-airport-express-playing/commit/66475cb0b75592fc9aab43149c663eb0921219ac))

## [1.3.0](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.2.0...v1.3.0) (2020-12-06)


### Features

* remove test script, it was dangerous ([eee6096](https://github.com/apexad/homebridge-airport-express-playing/commit/eee6096052cc4eb56c365e954f35d0e50ec1b30f))

## [1.2.0](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.1.3...v1.2.0) (2020-12-06)


### Features

* re-enable adding of all airport express and fix check for playing/not playing ([16bc5a6](https://github.com/apexad/homebridge-airport-express-playing/commit/16bc5a62d4a5b9dbcd3d73a0ec1560a9111780ce))

### [1.1.3](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.1.2...v1.1.3) (2020-12-06)


### Bug Fixes

* again try and correct test script ([f95804f](https://github.com/apexad/homebridge-airport-express-playing/commit/f95804ffbadcc50cab4de3ec4860c1675ef440a1))

### [1.1.2](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.1.1...v1.1.2) (2020-12-06)


### Bug Fixes

* convert test script to pure javascript ([040e626](https://github.com/apexad/homebridge-airport-express-playing/commit/040e626b59a34b67d2a46c39a224ebf592355422))

### [1.1.1](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.1.0...v1.1.1) (2020-12-06)


### Bug Fixes

* fix location of bin test script ([d71dca6](https://github.com/apexad/homebridge-airport-express-playing/commit/d71dca6542a6272fe94b769ca7491f89fc3fdbf4))

## [1.1.0](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.0.3...v1.1.0) (2020-12-06)


### Features

* add a testing script and some info about it readme ([c0059e9](https://github.com/apexad/homebridge-airport-express-playing/commit/c0059e99d4f9b7e730e1cdc720986fc979fc2a3d))

### [1.0.3](https://github.com/apexad/homebridge-airport-express-playing/compare/v1.0.2...v1.0.3) (2020-12-06)


### Bug Fixes


test.js
#!/usr/bin/env node
require("./dist/test.js");


src/test.ts
import mdns from 'mdns-js';
import { mDNSReply } from './settings';

mdns.excludeInterface('0.0.0.0')

const browser = mdns.createBrowser(mdns.tcp("airplay"));

browser.on('ready', () => {
  console.log('browser is ready');
  browser.discover();
});

browser.on('update', (data: mDNSReply) => {
  console.log('data:', data);
});

//stop after timeout
setTimeout(() => { browser.stop(); }, 5000);


src/platform.ts
import {
  API,
  DynamicPlatformPlugin,
  Logger,
  PlatformAccessory,
  PlatformConfig,
  Service,
  Characteristic,
} from 'homebridge';
import mdns from 'mdns-js';
import {
  mDNSReply,
  PLATFORM_NAME,
  PLUGIN_NAME,
} from './settings';
import AirportExpress from './platformAccessory';

export default class AirportExpressPlayingPlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;

  public readonly accessories: PlatformAccessory[] = [];

  constructor(
    public readonly log: Logger,
    public readonly config: PlatformConfig,
    public readonly api: API,
  ) {
    this.config = config;
    this.log.debug('Finished initializing platform:', this.config.name);

    this.api.on('didFinishLaunching', () => {
      log.debug('Executed didFinishLaunching callback');
      this.discoverDevices();
    });
  }

  configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName);
    this.accessories.push(accessory);
  }

  discoverDevices() {
    const mdnsBrowser = mdns.createBrowser(mdns.tcp("airplay"));

    mdnsBrowser.on('ready', () => {
      this.log.info('Searching for Airport Express devices')
      mdnsBrowser.discover();
    });
    
    mdnsBrowser.on('update', (data: mDNSReply) => {
      if (data && data.txt && data.txt.includes('model=AirPort10,115')) {
        const serialNumber = data.txt.find((str) => str.indexOf('serialNumber') > -1)?.replace('serialNumber=', '') || '';
        const displayName = data.fullname.replace('._airplay._tcp.local', '');
        const uuid = this.api.hap.uuid.generate(serialNumber);

        const existingAccessory = this.accessories.find(accessory => accessory.UUID === uuid);

        if (existingAccessory) {
          // the accessory already exists
          this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName);
          new AirportExpress(this, existingAccessory);
        } else {
          this.log.info('Adding new accessory:', displayName);

          const accessory = new this.api.platformAccessory(displayName, uuid);
          accessory.context.device = {
            serialNumber,
            displayName,
            data,
          };

          new AirportExpress(this, accessory);

          this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
        }
      }
    });

    setTimeout(() => mdnsBrowser.stop(), 5000);
  }
}

src/platformAccessory.ts
import {
  Service,
  PlatformAccessory,
  CharacteristicValue,
} from 'homebridge';
import mdns from 'mdns-js';
import AirportExpressPlayingPlatform from './platform';
import { mDNSReply } from './settings';

export default class ExamplePlatformAccessory {
  private service: Service;
  private readonly name: string;
  private readonly serialNumber: string;
  private readonly showSwitch: boolean;
  private readonly switchService!: Service;

  constructor(
    private readonly platform: AirportExpressPlayingPlatform,
    private readonly accessory: PlatformAccessory,
  ) {
    this.platform = platform;
    this.showSwitch = this.platform.config.showSwitch;
    this.name = accessory.context.device.displayName;
    this.serialNumber = accessory.context.device.serialNumber

    // set accessory information
    this.accessory.getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Apple Inc. via apexad')
      .setCharacteristic(this.platform.Characteristic.Model, 'AirPort10,115')
      .setCharacteristic(this.platform.Characteristic.SerialNumber, this.serialNumber);

    this.service = this.accessory.getService(this.platform.Service.SmartSpeaker) || this.accessory.addService(this.platform.Service.SmartSpeaker);
    this.service
      .setCharacteristic(this.platform.Characteristic.Name, this.name)
      .setCharacteristic(this.platform.Characteristic.ConfiguredName, this.name);

    if(this.showSwitch) {
      this.switchService = accessory.getService(this.platform.Service.Switch)
      || accessory.addService(this.platform.Service.Switch, `${this.name} Switch`, `${this.serialNumber} Switch`);
    }

    this.platform.log.info(`Airport Express device ${this.name} (serial number: ${this.serialNumber} created!`);

    this.setMediaState(this.convertMediaState(accessory.context.device.data.txt));
    setInterval(this.updateMediaState.bind(this), 5000);
  }

  updateMediaState() {
    this.platform.log.debug(`Updating Airport Exrpess with serial number ${this.serialNumber}`);
    const mdnsBrowser = mdns.createBrowser(mdns.tcp("airplay"));
    mdnsBrowser.on('ready', () => mdnsBrowser.discover());
    mdnsBrowser.on('update', (data: mDNSReply) => {
      try {
        if (data && data.txt) {
          const foundSerialNumber = data.txt.find((str) => str.indexOf('serialNumber') > -1)?.replace('serialNumber=', '');

          if (data.txt.includes('model=AirPort10,115') && foundSerialNumber && this.serialNumber === foundSerialNumber) {
            this.platform.log.debug(`txt record contents: ${data.txt}`)
            this.setMediaState(this.convertMediaState(data.txt));
            mdnsBrowser.stop();
          }
        }
      } catch(error) {
        this.platform.log.error(`Error in mDNS check, found invalid record`);
        this.platform.log.debug(error);
        mdnsBrowser.stop();
      }
      setTimeout(() => {
        try {
          // make sure mdnsBrowser was stopped if it was not stopped above
          mdnsBrowser.stop();
        } catch(err) {
          this.platform.log.debug(`mdns browser for stop via timeout error: ${err}`);
        }
      }, 5000);
    });
  }

  setMediaState(state: CharacteristicValue) {
    this.service
      .setCharacteristic(this.platform.Characteristic.TargetMediaState, state)
      .setCharacteristic(this.platform.Characteristic.CurrentMediaState, state);

    if (this.showSwitch) {
      this.switchService
        .setCharacteristic(
          this.platform.Characteristic.On,
          state === this.platform.Characteristic.CurrentMediaState.PLAY ? true : false  
        )
    }
  }

  convertMediaState(mDNS_TXT_record: Array<string>) {
    const bit11 = (parseInt(mDNS_TXT_record.find((r: string) => r.indexOf('flag') > -1)!.replace('flags=', ''), 16).toString(2)).padStart(12, '0').charAt(0);
    if (bit11 === '0') {
      return this.platform.Characteristic.CurrentMediaState.STOP;
    } else if (bit11 === '1') {
      /* bit11 correspponds to playing
       * see https://github.com/openairplay/airplay-spec/blob/master/src/status_flags.md
       */
      return this.platform.Characteristic.CurrentMediaState.PLAY;
    }
    return this.platform.Characteristic.CurrentMediaState.INTERRUPTED;
  }
}
